### DISCUSSION TITLE
Avoid reflecting wave in waveguide with gaussian source

### DISCUSSION THREAD

**Discussion by LaraCouronne** [NanoComp/meep#2436] [Q&A] [answered]
Hello !
I am new with meep. I am trying to simulate a waveguide in which propagates a gaussian source, in order to calculate the loss induced by the shape and material of that waveguide. 
However there is an issue : when I animate the propagation of the source, I can see that the wave is reflected despite the PML boundary conditions, which changes the fluxes values.

To understand what was wrong with my code, I tried looking at examples from the documentation where they calculate exactly what I am trying to obtain. Here is an exact code from the documentation [examples/bend-flux.py](https://github.com/NanoComp/meep/blob/master/python/examples/bend-flux.py) in which I just animated the geometry to check if the same effect occured :

```
import meep as mp

resolution = 10  # pixels/um

sx = 16  # size of cell in X direction
sy = 32  # size of cell in Y direction
cell = mp.Vector3(sx, sy, 0)

dpml = 1.0
pml_layers = [mp.PML(dpml)]

pad = 4  # padding distance between waveguide and cell edge
w = 1  # width of waveguide

wvg_xcen = 0.5 * (sx - w - 2 * pad)  # x center of vert. wvg
wvg_ycen = -0.5 * (sy - w - 2 * pad)  # y center of horiz. wvg

geometry = [
    mp.Block(
        size=mp.Vector3(mp.inf, w, mp.inf),
        center=mp.Vector3(0, wvg_ycen, 0),
        material=mp.Medium(epsilon=12),
    )
]

fcen = 0.15  # pulse center frequency
df = 0.1  # pulse width (in frequency)
sources = [
    mp.Source(
        mp.GaussianSource(fcen, fwidth=df),
        component=mp.Ez,
        center=mp.Vector3(-0.5 * sx + dpml, wvg_ycen, 0),
        size=mp.Vector3(0, w, 0),
    )
]

sim = mp.Simulation(
    cell_size=cell,
    boundary_layers=pml_layers,
    geometry=geometry,
    sources=sources,
    resolution=resolution,
)


pt = mp.Vector3(0.5 * sx - dpml - 0.5, wvg_ycen)

animate = mp.Animate2D(fields=mp.Ez,
                       realtime=True,
                       field_parameters={'alpha': 0.8, 'cmap': 'RdBu', 'interpolation': 'none'},
                       boundary_parameters={'linewidth': 1.5, 'facecolor': 'y', 'edgecolor': 'b',
                                            'alpha': 0.3})

sim.run(mp.at_every(1, animate), until_after_sources=mp.stop_when_fields_decayed(50, mp.Ez, pt, 1e-3))
```

We can see that there is a reflection of the wave coming back in the waveguide, even in that simple code from the documentation.
So my first question is : **is there a way to avoid that ?**

I thought about changing the simulation time to make it stops before the wave comes back, which leads to my second question :
**Is there a way to calculate at which timestep that reflection phenomenon will occur, to know when to stop the simulation ?**

I am very confused by that phenomenon, I thought that the PML boundary conditions where here to avoid that kind of effet.

Regards,
Lara Couronné

**Comment by Andeloth** [comment, marked as answer]
You haven't normalized the simulation. Once normalized and converted to a gif/mp4 at the end of the simulation by setting `normalize=True` in the Animate2D constructor, you get the following animation:

https://user-images.githubusercontent.com/37753156/224758509-a39016f4-bfca-469d-b431-460277f8ad54.mp4

It seems like there is a very small reflection in the system due to the PML which is a numerical artifact and is likely unavoidable, but it is not very significant due to the magnitude of the reflected fields. I believe there is a way to record the amount of reflected flux to measure how much is reflected, but I don't remember how off the top of my head.

**Reply by LaraCouronne** [reply]
I'll look up the documentation to find how to get and substract the amount of reflected flux then, but you're right, it's significantly smaller than what I thought. Thank you so much !

**Comment by stevengj** [comment]
I'm not sure what you mean by "reflection".  I see a lot of radiation in your simulation, but that's because you are using a constant spatial profile and not an eigenmode source, so of course your source radiates power into the cladding as well as coupling to the waveguide.

(Of course, there is always a small reflection from the PML too, but I can't see it with my eyes in the animation you posted.  The PML reflection goes to zero with increasing resolution or increasing PML thickness, and is usually quite small to begin with.)

(An example showing how to explicitly isolate the reflected wave in a structure is given in the [bend transmission tutorial](https://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/#transmittance-spectrum-of-a-waveguide-bend).)

**Reply by Andeloth** [reply]
Just for reference, here is the non-normalized animation, you can see the reflected wave around 4 seconds in.

https://user-images.githubusercontent.com/37753156/224846736-c7ce2a1f-8af1-4d2c-89ab-9fd1cfbcc911.mp4

**Reply by Andeloth** [reply]
It's just a very small reflection which can be likely decreased with increased resolution and PML width, as you said.

**Reply by stevengj** [reply]
Ah, I see.  Yes, in the non-normalized animation you are effectively *exponentially amplifying* the image (thanks to the exponential decay of the original pulse), so even if the reflection is `1e-8` you would still see it.
